merge split_naive and split_constr: a simpler recursion scheme

trefis opened the pull request:
Aka: *No! **You** review*.

Remember the docstring at the top of the file that you so confidently approved:
```
  (split_and_precompile)                                                    
We first split the initial pattern matching (or "pm") along its first column
-- simplifying pattern heads in the process --, so that we obtain an ordered
list of pms.                                                                
For every pm in this list, and any two patterns in its first column, either 
the patterns have the same head, or they match disjoint sets of values. (In 
particular, two extension constructors that may or may not be equal due to  
hidden rebinding cannot occur in the same simple pm.)                       
```

We had some back and forth about the second paragraph of that one, and I think we are going to need to refine it further.
While I believe it is correct (the pms do satisfy this invariant), it is actually weaker than the post-condition provided by `split_and_precompile`, indeed its current wording currently allows the following to lines in the same pm:
- `(Ext_C1, true)` and `(Ext_C2, false)` even when `Ext_C1 ≟ Ext_C2`
- `(C1, true)` and `(_, false)`

The post-condition is actually:
```
∀pm, ∀row1 row2 in pm,
  kind(head(row1)) = kind(head(row2))
  && (head(row1) = head(row2) || head(row1) ≠ head(row2))
```
Note:
- until the addition of extensible types, the second part trivially held
- I will let you put that into words so that we can update the comment at the top of the file

If we look at `split_constr` and `split_naive`, they both make the assumption that there can be only two kinds of patterns on the first column: omegas, or discriminating patterns. Giving birth to the recursion scheme based on `split_ex / split_noex` (that we renamed to `collect_group / collect_var`).

I believe we can ignore that assumption, and produce a much simpler scheme: we have a single `collect` function, which goes through clauses in order; for each clause the **important** question is this: can we add it to the group/pm that we are currently producing? That is: does it very the condition stated above?
If yes, we add it to the group and we keep going, if not then there is a follow up *less important* question: do we split or do we try to look at the following rows and try to raise them? (It is less important because it does not affect correctness)
The heuristic used to answer is the one discussed will depend on the kind of the group being built (is its first column populated by extension constructors?).

I think the result is quite a lot simpler than before, hopefully you'll like it.

trefis committed:
matching: merge split_naive and split_constr

trefis committed:
move things around

trefis committed:
highlight similarity between collect_group and collect_vars

trefis committed:
this was incorrect: the first line is no is not necessarily a var, it
could also be an different extension constructor. let's not try to be
smart and just always go through [split].

trefis committed:
share splitting code between both collect functions

trefis committed:
get_group => can_group

trefis committed:
merge collect_group and collect_var

trefis committed:
right, this should indeed work as before

trefis committed:
to split or not to split

trefis commented on bytecomp/matching.ml:
```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
       else
-        split_noexc [ cl ] rem
-  | _ -> assert false
+        precompile_normal
```
Note: we can make the above disappear if we rewrite the code below slightly; in case you don't like it.

trefis commented:
PS: for once I think it makes as much sense to review the whole diff as it does to review by commit. Feel free to choose the approach you prefer!

gasche commented:
Note on the explanation comment:

> either the patterns have the same head, or they match disjoint sets of values

maybe we should rephrase as

> either the patterns have the same head, or their heads match disjoint sets of values

gasche commented on bytecomp/matching.ml:
```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only *)
+        collect group_discr rev_yes (cl :: rev_no) []
```
do we know anything about what this special case does in the non-variable case?

gasche replied:
(you could avoid the question by checking `group_var group_discr` as an extra condition)

gasche replied:
(@trefis replied in the main discussion that, in the non-variable case, this creates a different group anyway.)

gasche commented on bytecomp/matching.ml:
```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
+             Otherwise, we try to raise lower clauses. *)
+          match group_discr.pat_desc with
+          | Tpat_construct (_, { cstr_tag = Cstr_extension _ }, _) ->
+              assert (rev_no = []);
```
I'm still not fond of the way split-global coherence assumptions are interleaved with clause-local decision-making here, resulting in this assertion.

gasche reviewed (commented):
We might be hitting diminishing returns.

I like `precompile_normal` -- I think it could be called `precompile_constr`, with a comment saying that no actual precompilation actually happens in this case (because the current implementation looks weirdly short).

Removing `collect_var` as a special case may be a good move, but then the weird optimisation case pops up in the common function, which feels a bit like we've added the complexity of two distinct special cases to a generic function.

Finally we still have this unpleasant way that some global invariant (we split at all patterns incompatible with an extension constructor) implies a local assertion (we haven't looked below any incompatible patterns, so `rev_no` is empty), that I don't like. (But what is a matter of taste and what is objective?)

I think that a personal favorite would be a merge of my proposal and this one, where I keep my three distinct  `collect_*` functions (they contain different special cases, and their interface is not uniform), but I keep your generalized notion of precompilation to clarify how the "end" of each collection is the same -- and share the code.

trefis commented:
> We might be hitting diminishing returns.

I respectfully disagree. I think this version is much better than the 1st version, and that both 2nd versions (mine and yours).

> I like `precompile_normal` -- I think it could be called `precompile_constr`, with a comment saying that no actual precompilation actually happens in this case (because the current implementation looks weirdly short).

Yes, `precompile_normal` is a bad name.
I'm ~OK with `precompile_constr`.

> Removing `collect_var` as a special case may be a good move, but then the weird optimisation case pops up in the common function, which feels a bit like we've added the complexity of two distinct special cases to a generic function.

> do we know anything about what this special case does in the non-variable case? (you could avoid the question by checking group_var group_discr as an extra condition)

The special case doesn't change a thing in the non-variable case: the clause is of a different kind than the "group", and so would have ended up in `no` anyway.

> Finally we still have this unpleasant way that some global invariant (we split at all patterns incompatible with an extension constructor) implies a local assertion (we haven't looked below any incompatible patterns, so rev_no is empty), that I don't like. (But what is a matter of taste and what is objective?)

I get your point. Note that we can replace:
```ocaml
assert (rev_no = []);                         
let yes = List.rev rev_yes in                 
insert_split group_discr yes (cl :: rem) def k
```
with
```ocaml
let yes = List.rev rev_yes in                 
insert_split group_discr yes (List.rev_append (cl :: no) rem) def k
```
Which I find more aesthetically pleasing, and it's also easier to reason about the correctness of the function with that version.
However, I felt like it didn't give as much information about what's going on (but perhaps that information should be in a comment? Or doesn't really matter?).

> I think that a personal favorite would be a merge of my proposal and this one, where I keep my three distinct collect_* functions

I'm sorry but that's a big no from me.
I am indeed slightly sad about the special case for the last row, but apart from that I think having one simple loop instead of 3 interleaved is much simpler.

trefis commented:
> Yes, `precompile_normal` is a bad name.
> I'm ~OK with `precompile_constr`.

Actually I just noticed that this function is *exactly* `dont_precompile_var` which already exists. How about we merge the two into `don't_precompile`?

gasche commented:
`do_not_precompile`, yes, works for me.

I wondered about the `List.rev_append` version myself. I also think it's not better.

It looks like we don't agree on which approach is the clearest one. (Here I prefer to have two or three clear loops instead of one complex one, even at the cost of some redundancy. I would not be surprised if this was also Luc's style, btw.) I propose to go with your approach, since you wrote the code.

gasche commented on bytecomp/matching.ml:
```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only *)
+        collect group_discr rev_yes (cl :: rev_no) []
```
I maybe have an idea of what this optimization is good for (as far as I know, we still don't understand this point). If we did *not* split here, then the whole group would have this catch-all row in its default-environment matrix, and it would never disappear due to further specialization. With the split here, then the first part of the group may be specialized away at some point, and the one-row group that we are creating would remain alone in the default environment; pattern matchings that fail earlier on would jump to it directly, which is more efficient than going through the first part of the matrix first.

gasche replied:
Note: the patterns in the tail of `ps` are not in simple form, so `group_var` is not the right function to call here -- in particular, it does not accept `Tpat_var` which is likely to occur. It would make sense to call `fun p -> le_pat p omega` -- which is also constant-time.

trefis replied:
> maybe have an idea of what this optimization is good for ...

Yes, I had reached the same conclusion the other day. It is in fact fairly easy to write a test case demonstrating it:
```ocaml
let f = function
  | true, false -> 1
  | _, false -> 2
  | _, _ -> 3
```
On trunk, 4.08, and this branch the output of `ocamlc -dlambda -c -nostdlib -nopervasives ./test.ml` is:
```
(setglobal Test!
  (let
    (f/3 =
       (function param/4 : int
         (catch
           (if (!= (field 0 param/4) 0)
             (if (!= (field 1 param/4) 0) (exit 1) 1)
             (if (!= (field 1 param/4) 0) (exit 1) 2))
          with (1) 3)))
    (makeblock 0 f/3)))
```
Now, if I comment out the optimisation clause, it changes to the following:
```
(setglobal Test!
  (let
    (f/3 =
       (function param/4 : int
         (catch
           (if (!= (field 0 param/4) 0)
             (if (!= (field 1 param/4) 0) (exit 1) 1) (exit 1))
          with (1) (if (!= (field 1 param/4) 0) 3 2))))
    (makeblock 0 f/3)))
```
Should I add this to the testsuite and add a comment in the code referring to it?

> Note: the patterns in the tail of ps are not in simple form, so group_var is not the right function to call here -- in particular, it does not accept Tpat_var which is likely to occur. It would make sense to call fun p -> le_pat p omega -- which is also constant-time.

Good catch indeed.
However, I can't help but notice that this is already the case in the original version, could it be on purpose?

trefis replied:
> It would make sense to call `fun p -> le_pat p omega` -- which is also constant-time.

I think that's wrong actually, let's say `p` is `C (D (E _))`. `le_pat p omega` ends up calling `not (satisfiable [[p]] [omega])`, and `satisfiable` will end up recursing a a few times (doing a bunch of non-`O(1)` operations each time).

I vote for not touching this for now.

trefis committed:
matching: dedup precompile_normal and dont_precompile_var

trefis committed:
matching: comment split_no_or

trefis commented:
Submitted upstream.

gasche commented:
Re. the optimization: I think it would be nice to (0) have a comment explaining this at the point of optimization (I haven't read the upstream PR yet to check whether you did so), (1) add the example you describe to the testsuite (with the -dno-unique-ids lambda output?), (2) look at the code produced in a case where variables are used to capture the output (and the difference if we optimize the ariable case) and (3) if we don't use `le_pat`, at least consider a non-simple version of `group_var` that traverses variables and aliases (and, if we want to be fancy, does an "or" on or-patterns).

gasche commented:
Your remark on `le_pat` makes me wonder whether it was wise ot use it in the or-simplification function. Maybe we should on the contrary assume that we compile code where the "useless subpattern" warning doesn't fire, and remove that simplification. (Assuming all cases that are simplifiable are indeed of that form. But then maybe with pattern synonyms we get more cases where simplifications are beneficial?)

trefis commented:
> I think it would be nice to (0) have a comment explaining this at the point of optimization (I haven't read the upstream PR yet to check whether you did so), (1) add the example you describe to the testsuite (with the -dno-unique-ids lambda output?), (2) look at the code produced in a case where variables are used to capture the output (and the difference if we optimize the ariable case) 

I merged these three: the comment points to an example of it applying, and one of it not applying (because we have variables and not wildcards on the last line), I think it's enough to show what's going on.
I did not use `-dno-unique-ids` because that makes the output less readable (every variable is either `*match*` or `*param*`), but since this is built with `-nostdlib -nopervasives` the tests should be relatively stable.

> (3) if we don't use le_pat, at least consider a non-simple version of group_var that traverses variables and aliases (and, if we want to be fancy, does an "or" on or-patterns).

Let's put a pin in that one.

trefis commented:
> Your remark on le_pat makes me wonder whether it was wise ot use it in the or-simplification function. Maybe we should on the contrary assume that we compile code where the "useless subpattern" warning doesn't fire, and remove that simplification. (Assuming all cases that are simplifiable are indeed of that form. But then maybe with pattern synonyms we get more cases where simplifications are beneficial?)

Yeah, I agree it's somewhat worrisome.
I'm not sure I understand your point about pattern synonyms.
